### DISCUSSION TITLE
NavigationSplitView support?

### DISCUSSION THREAD

**Discussion by Alexminator99** [pointfreeco/swift-composable-architecture#2894] [Q&A] [answered]
Hi there community. Im having some troubles to integrate NavigationSplitView to a TCA aligned arch. What I tried so far is one link on [stack overflow](https://stackoverflow.com/questions/76994708/using-navigationsplitview-with-the-composable-architecture).

The thing is that he uses the same destination but in my case I have 2 different views. On one side you have 2 possible destinations, so I applied somehow what the link described and got it working but I'm experiencing some issues of course with that approach. 

Is there an example that approaches this the TCA way? At least the minimum essential to get started.

## System
- Xcode 15.2
- Swift 5.9
- iOS 17

**Comment by mbrandonw** [comment]
Hi @Alexminator99, I am not aware of any TCA-specific problems having to do with `NavigationSplitView`. You should first ask yourself if you know how to accomplish what you want using vanilla SwiftUI? And then you should ask can you adapt the vanilla SwiftUI code to work with your TCA features?

If you do know how to accomplish this in vanilla SwiftUI then perhaps you can provide a minimal, compiling example of it using only the tools from SwiftUI. Then someone can help you adapt it to TCA.

**Reply by Alexminator99** [reply]
Hi @mbrandonw the issue is a little more complex tbh. The app uses a sidebar at the top right that is independent of the sidebar of the `NavigationSplitView`. Im using a mutually exclusive approach for navigating through the screens in the `NavigationSplitView` (the sidebar). That logic works like a charm, the problem relies on the detail part (I do not have a content part), because for example, lets assume we open a View that have a .task (Hardware) that is cancellable and using Tree navigation you go to another view (Clover). So far so good, but if the user decides to go to the Main Sidebar (the top right) and navigate to lets say Home, the CloverView is properly dismissed and disposed but Hardware's `.task` gets executed and I have a warning/error from the library of an unexpected action.

![CleanShot 2024-03-14 at 08 49 30@2x](https://github.com/pointfreeco/swift-composable-architecture/assets/51393761/920e1ab8-9559-4cae-bbba-bc846fa9af98)

My question is how I properly dispose ViewA in the NavigationSplitView that is mutually exclusive (in order to have only 1 screen displayed, I used Tic-Tac-Toe as guide for that). Sorry for the original post was not descriptive as this. I will edit this comment and add the code later for its analysis. 

![CleanShot 2024-03-14 at 08 53 47@2x](https://github.com/pointfreeco/swift-composable-architecture/assets/51393761/9cfd896c-2ddb-4141-943f-8647937e9278)

As final details, the image is self explanatory, there is a parent Reducer that handles that sidebar. Normally there wouldn't be any issue if the `NavigationSplitView` were the main view but is a little more complex this time. Anyway, if you need an example, I will share the parts of the code we need to check.

**Reply by mbrandonw** [reply]
Hi @Alexminator99, I'm sorry but I don't personally have time to figure out what you are trying to accomplish form your description above. Can you provide a vanilla SwiftUI project that demonstrates what you are trying to achieve, and then maybe someone can help you adapt it to TCA?

And if you can't create a vanilla SwiftUI version of what you want then it sounds like this isn't really a TCA-specific question/issue, but rather just a general question about SwiftUI, in which case I think you will get more help in other discussion forums out there.

**Reply by Alexminator99** [reply]
> Hi @Alexminator99, I'm sorry but I don't personally have time to figure out what you are trying to accomplish form your description above. Can you provide a vanilla SwiftUI project that demonstrates what you are trying to achieve, and then maybe someone can help you adapt it to TCA?
> 
> 
> 
> And if you can't create a vanilla SwiftUI version of what you want then it sounds like this isn't really a TCA-specific question/issue, but rather just a general question about SwiftUI, in which case I think you will get more help in other discussion forums out there.

I understand. I will create a repo thks so anyone can replicate the behavior.

**Comment by robfeldmann** [comment, marked as answer]
Hi @Alexminator99, I have found this example very helpful in my own split view setup with TCA: https://github.com/kodydeda4/TCA-NavigationSplitView

**Reply by Alexminator99** [reply]
Hi @robfeldmann thanks for the example, I will take a look. I added a comment with a lot more information than the original question. I appreciate your time 🙌

**Reply by robfeldmann** [reply]
I saw your updates. Sorry, I'm always on mobile when I have time to check discussions. If I recall correctly, the example I linked to has presentation state for the sidebar which effects what content is shown in the detail. But then whatever is in the detail column has its own presentation state (for alerts, dialogs, sheets, whatever). My gut says you will need a pattern like this somehow.

In your example it is unclear to me how the right sidebar thing is structured. Does the detail own it (almost like an inspector column) and when something in that menu is selected, is the content supposed to show in the detail? 

Just some things to think about to help you figure out the best way to model this.

**Reply by Alexminator99** [reply]
I will provide tonight an example of what I have exactly an MVP so is clear what's the goal I want to accomplish and I don't know if is me who doesn't know how to get it or if it is an edge case that is not covered yet by TCA. I will ping you if you want to take a look and check :). Thks again...

**Comment by JozefLipovsky** [comment]
I recently started exploring TCA and ran into a [problem](https://pointfreecommunity.slack.com/archives/C04KQQ7NXHV/p1710387684973279) with understanding how to use NavigationSplitView in TCA. I'm going through all the discussions and checking those links right now. I really appreciate how many people are actively contributing to discussions and PRs here and on Slack. This is super helpful 🙌

**Comment by Alexminator99** [comment]
> Hi @Alexminator99, I have found this example very helpful in my own split view setup with TCA: https://github.com/kodydeda4/TCA-NavigationSplitView

This solution migrated to 1.9 works like a charm. Thanks a lot my friend!!!.
